### ISSUE TITLE
Add derive support for enums via `UnionArray`

### ISSUE THREAD

**Issue by mbrobbel** [mbrobbel/narrow#90]
Allow deriving `ArrayType` for sum types via `UnionArray`.

**Comment by mbrobbel** [comment]
An idea, given an enum like:

```rust
#[derive(Clone)]
enum Foo {
    Unit,
    One(u32),
    Unnamed(u8, u16),
    Named {
        a: bool,
        b: String,
    },
}
```

And the following helper trait:
```rust
/// Using `i8` to match the arrow spec for union variant indices (types).
trait EnumVariant<const INDEX: i8>: Sized {
  /// The data for this variant, this needs `ArrayType` to 
  /// allow storing instaces of these in arrays.
  type Data: ArrayType;

  /// Given the data of this variant, return the variant of the
  /// original enum (`Self`).
  fn from_data(data: Self::Data) -> Self;
}
```

Then generate something along the lines of:
```rust
// we can choose to add a `Unit` type for the data of this variant
struct FooUnit;
unsafe impl Unit for FooUnit {
  type Item = Foo;
}
// we need this conversion for the `Unit` trait and `NullArray` index and iterators implementations
// to return instances of Foo
impl From<FooUnit> for Foo {
  fn from(_value: FooUnit) -> Self {
    Foo::Foo
  }
}

impl EnumVariant<0> for Foo {
  // and then use that here,
  type Data = FooUnit;
  // Or we use (), this means we get a `NullArray<()>` (because `<() as ArrayType>::Array` is `NullArray`)
  // meaning we have to convert items from this array elsewhere (non-ideal)
  type Data = ();

  fn from_data(_data: Self::Data) -> Self {
    Foo::Unit
  }
}

impl EnumVariant<1> for Foo {
  // we can directly use the one field here
  type Data = u32;
  
  fn from_data(data: Self::Data) -> Self {
    Foo::One(data)
  }
}

// for multiple unnamed fields we can derive an array type
#[derive(Copy, Clone, ArrayType)]
struct FooUnnamed(u8, u16);

impl EnumVariant<2> for Foo {
  type Data = FooUnnamed;

  fn from_data(data: Self::Data) -> Self {
    Foo::Unnamed(data.0, data.1)
  }
}

// for named values we also derive a struct array type
#[derive(Copy, Clone, ArrayType)]
struct FooNamed {
};
```

**Comment by mbrobbel** [comment]
Closed via #146.
